sentry-core/src/session.rs

```diff
@@ -169,10 +178,24 @@ impl SessionFlusher {
         }
     }
 
+    /// Enqueues a session update for delayed sending.
+    ///
+    /// When the queue is full, it will be flushed immediately.
     pub fn enqueue(&self, session_update: SessionUpdate<'static>) {
-        self.queue.lock().unwrap().push(session_update)
+        let flush_immediately = {
+            let mut queue = self.queue.lock().unwrap();
+            queue.push(session_update);
+            queue.len() >= MAX_SESSION_ITEMS
+        };
+        if flush_immediately {
+            SessionFlusher::flush(&self.queue, &self.transport);
```

Comment:
This is potentially racy. Between checking the queue length and calling `flush`, another thread could enqueue and flush, or a flush could happen. 

To avoid this, consider keeping the lock and passing a lock guard into `flush`.